### PR TITLE
Add sender HMAC to MessageV2

### PR DESCRIPTION
**Pull request by nplasterer**
Closes https://github.com/xmtp/xmtp-ios/issues/216

Follows the pattern put in place here: https://github.com/xmtp/xmtp-js/pull/519

This fixes the iOS entitlement issue for push notifications by adding additional information to the payload. See: https://github.com/xmtp/XIPs/pull/35

### CODE REVIEW

File: Sources/XMTP/Codecs/ReplyCodec.swift
Comment (nakajima) [inline Sources/XMTP/Codecs/ReplyCodec.swift:71]: Do we not want to push for replies?

File: Sources/XMTP/Codecs/RemoteAttachmentCodec.swift
Comment (nakajima) [inline Sources/XMTP/Codecs/RemoteAttachmentCodec.swift:211]: Do we not want to push for remote attachments?

File: Tests/XMTPTests/MessageTests.swift
Comment (nakajima) [inline Tests/XMTPTests/MessageTests.swift:47]: Should we bring this back?

File: Sources/XMTP/Codecs/ReplyCodec.swift
Comment (kele-leanes) [reply Sources/XMTP/Codecs/ReplyCodec.swift:71]: the scope of the task only includes `textCodec` and `Composite`

File: Sources/XMTP/Codecs/RemoteAttachmentCodec.swift
Comment (kele-leanes) [reply Sources/XMTP/Codecs/RemoteAttachmentCodec.swift:211]: the scope of the task only includes `textCodec` and `Composite`

Comment (nplasterer) [reply Sources/XMTP/Codecs/RemoteAttachmentCodec.swift:211]: 👋 @kele-leanes I actually shared a table with you that showed the rest of the codecs as well. We definitely want this to be true. Just because JS has codecs in a different repo doesnt mean they aren't also supported. Here is the table for reference again.

Screenshot 2024-01-30 at 7 26 54 AM

File: Tests/XMTPTests/CodecTests.swift
Comment (nplasterer) [inline Tests/XMTPTests/CodecTests.swift:12]: Would you mind adding a test in here that shows you can get the shouldPush method correctly from the decodedMessage kind of like we should have fallback tests. 🙏

**Review by nplasterer — COMMENTED** [review]
Looks like you have two legit linter issues

`/Users/runner/work/xmtp-ios/xmtp-ios/Sources/XMTPiOS/Messages/MessageV2.swift:105:91: error: Force Unwrapping Violation: Force unwrapping should be avoided `

`/Users/runner/work/xmtp-ios/xmtp-ios/Sources/XMTPiOS/ConversationV2.swift:249:50: error: Force Cast Violation: Force casts should be avoided (force_cast)`

**Comment by nplasterer** [comment]
I can't approve this since I opened it but would be good to get eyes from @rygine and or @Bren2010 on the cryptography parts and @nakajima on the iOS code 🙏  But looks good to me.

**Comment by rygine** [comment]
the crypto code _seems_ OK, but i'll defer to someone with more knowledge. all the other pieces look to be in the right places.

File: Tests/XMTPTests/CodecTests.swift
Comment (kele-leanes) [reply Tests/XMTPTests/CodecTests.swift:12]: seems that `Xmtp_MessageContents_EncodedContent` does not have the `shouldPush` method available which is where the fallback is coming from. Should I get it from there or from a different place?

Comment (nplasterer) [reply Tests/XMTPTests/CodecTests.swift:12]: That was my bad from looking at JS it's on the messagev2 so wrote a test for that scenario and pushed it up. https://github.com/xmtp/xmtp-ios/pull/218/commits/1967934ae85789b62885e2850de6e008840eb97b

**Review by nakajima — APPROVED** [review]

**Comment by nplasterer** [comment]
@kele-leanes do you mind bumping the pod spec so we can consume this in RN?

File: Tests/XMTPTests/CodecTests.swift
Comment (kele-leanes) [reply Tests/XMTPTests/CodecTests.swift:12]: thanks! @nplasterer

**Pull request merged**
